stdlib/socket/src/main/java/org/ballerinalang/stdlib/socket/compiler/SocketCompilerPlugin.java

```diff
@@ -0,0 +1,173 @@
+import static org.ballerinalang.model.types.TypeKind.OBJECT;
+import static org.ballerinalang.model.types.TypeKind.RECORD;
+import static org.ballerinalang.stdlib.socket.SocketConstants.RESOURCE_ON_ACCEPT;
+import static org.ballerinalang.stdlib.socket.SocketConstants.RESOURCE_ON_CLOSE;
+import static org.ballerinalang.stdlib.socket.SocketConstants.RESOURCE_ON_CONNECT;
+import static org.ballerinalang.stdlib.socket.SocketConstants.RESOURCE_ON_ERROR;
+import static org.ballerinalang.stdlib.socket.SocketConstants.RESOURCE_ON_READ_READY;
+import static org.ballerinalang.util.diagnostic.Diagnostic.Kind.ERROR;
+
+/**
+ * Compiler plugin for validating Socket services.
+ *
+ * @since 0.985.0
+ */
+@SupportedResourceParamTypes(
+        expectedListenerType = @SupportedResourceParamTypes.Type(packageName = "socket", name = "Listener"),
+        paramTypes = { @SupportedResourceParamTypes.Type(packageName = "socket", name = "Client") }
+)
+public class SocketCompilerPlugin extends AbstractCompilerPlugin {
+
+    private static final String INVALID_RESOURCE_SIGNATURE = "Invalid resource signature for %s in service %s. ";
+    private DiagnosticLog diagnosticLog = null;
+    private int resourceCount = 0;
+
+    @Override
+    public void init(DiagnosticLog diagnosticLog) {
+        this.diagnosticLog = diagnosticLog;
+    }
+
+    @Override
+    public void process(ServiceNode serviceNode, List<AnnotationAttachmentNode> annotations) {
+        List<BLangFunction> resources = (List<BLangFunction>) serviceNode.getResources();
+        resources.forEach(res -> validate(serviceNode.getName().getValue(), res, this.diagnosticLog));
+//        if (resourceCount != 4) {
+//            String msg = String
+//                    .format("Service needs to have all 4 resources[(%s or %s), %s, %s, %s].", RESOURCE_ON_ACCEPT,
+//                            RESOURCE_ON_CONNECT, RESOURCE_ON_READ_READY, RESOURCE_ON_CLOSE, RESOURCE_ON_ERROR);
+//            diagnosticLog.logDiagnostic(ERROR, serviceNode.getPosition(), msg);
+//        }
+    }
+
+    private void validate(String serviceName, BLangFunction resource, DiagnosticLog diagnosticLog) {
+        switch (resource.getName().getValue()) {
+            case RESOURCE_ON_CONNECT:
+            case RESOURCE_ON_ACCEPT:
+                validateOnAccept(serviceName, resource, diagnosticLog);
+                resourceCount++;
+                break;
+            case RESOURCE_ON_READ_READY:
+                validateOnReadReady(serviceName, resource, diagnosticLog);
+                resourceCount++;
+                break;
+            case RESOURCE_ON_CLOSE:
+                validateOnClose(serviceName, resource, diagnosticLog);
+                resourceCount++;
+                break;
+            case RESOURCE_ON_ERROR:
+                validateOnError(serviceName, resource, diagnosticLog);
+                resourceCount++;
+                break;
+            default:
```

Comment:
If the default case isn't supposed to be hit, better throw an error. That would avoid the possibility of someone adding another state and not doing the relevant change here. For all we know, adding another state could even be out of the logic.

Reply:
We can't throw errors. There can be other functions that mix with above mandatory resources which is a valid use case. We need to make sure Listener or client's callbackService has the above functions.